src/ngx_http_lua_shdict.c

```diff
@@ -1515,14 +1558,14 @@ ngx_http_lua_shared_dict_get(ngx_shm_zone_t *zone, u_char *key_data,
 
     ctx = zone->data;
 
-    ngx_shmtx_lock(&ctx->shpool->mutex);
+    ngx_http_lua_shdict_lock(ctx, NGX_HTTP_LUA_SHDICT_RLOCK);
```

Comment:
This looks wrong to me. You know that the `ngx_http_lua_shdict_unlock()` call is not really read only? It also always updates the LRU queue, which is not safe when multiple readers trying to change it at the same time. We need finer grained locking here.

Reply:
I think this comment is correct in a number of places, but not in `ngx_http_lua_shdict_get_keys` in particular because there is no LRU queue adjustment that I can see.

Reply:
@p0pr0ck5 True. Wrong place.